rfcs/20190729-keras-preprocessing.md

```diff
@@ -0,0 +1,738 @@
+            If set to False, sorts the data in alphanumeric order.
+        seed: Optional random seed for shuffling and transformations.
+        follow_links: Whether to follow links inside
+            subdirectories (default: False).
+        validation_split: Optional float between 0 and 1,
+            fraction of data to reserve for validation.
+        subset: One of "training" or "validation". Only used if `validation_split` is set.
+    """
+
+def dataset_from_directory(
+    self,
+    directory,
+    targets='inferred',
+    target_mode='categorical',
+    class_names='inferred',
+    color_mode='rgb',
+    batch_size=32,
+    shuffle=True,
+    seed=None,
+    follow_links=False,
+    validation_split=None,
+    subset='training'):
+    """Takes a directory and returns a Dataset yielding batches of images.
+
+    # Arguments:
+        directory: Directory where the data is located.
+            If `targets` is "inferred", it should contain
+            subdirectories, each containing images for a class.
+            Otherwise, the directory structure is ignored.
+        targets: Either
+            "inferred" (targets are generated from the directory structure),
+            None (no targets),
+            or a list of integer labels of the same size as the number of image
+            files found in the directory.
+        target_mode:
+            - 'categorical' means that the inferred labels are
+                encoded as a categorical vector (e.g. for categorical_crossentropy).
+            - 'binary' means that the inferred labels (there can be only 2)
+                are encoded as binary scalars (e.g. for binary_crossentropy).
+        class_names: Only valid if "targets" is "inferred". This is the explict
+            list of class names (must match names of subdirectories). Used
+            to control the order of the classes (otherwise alphanumerical order is used).
+        color_mode: One of "grayscale", "rgb", "rgba". Default: "rgb".
+            Whether the images will be converted to
+            have 1, 3, or 4 channels.
+        batch_size: Size of the batches of data (default: 32).
+        shuffle: Whether to shuffle the data (default: True)
+            If set to False, sorts the data in alphanumeric order.
+        seed: Optional random seed for shuffling and transformations.
+        follow_links: Whether to follow links inside
+            subdirectories (default: False).
+        validation_split: Optional float between 0 and 1,
+            fraction of data to reserve for validation.
+        subset: One of "training" or "validation". Only used if `validation_split` is set.
+    """
+```
+
+Notes:
+- Resizing (`target_size`) is moved to constructor (`resize`).
+- `validation_split` is moved from constructor to here.
+
+#### Question: how to support image segmentation in a simple way?
+
+**Requirements:**
+- Image loading and image augmentation should be synced across inputs and targets
```

Comment:
Keeping things synced is simple. Once the user passes the `inputs` and `targets`, we should store the `input filename` and `target filename` in a pandas dataframe. 

```
         inputfile            targetfile
0      path_to_file1     path_to_target1
1      path_to_file2     path_to_target2
...       ...                      ...
``` 
1) In order to shuffle the data and keeping `inputs` and `targets` synced, we only have to shuffle the indices of this `dataframe`. 
2) Any augmentation that is applied on a particular `input` will be applied on the `target` as well because they can be processed by the same thread in a sequential manner now.

Reply:
It's almost what flow_from_dataframe already does. So maybe we can beef up this function to perform whatever we need.